Added PodSecurityPolicy support required by istio sidecar

yu2003w opened the pull request:
Signed-off-by: Wei Yu(Jared) <yuweiw@cn.ibm.com>

<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Fixes #4480 

## Proposed Changes

* Added PodSecurityPolicy support for activator and autoscaler as Istio sidecar need capabilities "NET_ADMIN" to start.
This PR partially fix #4480.

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
None
```

markusthoemmes commented on config/204-psp.yaml:
```diff
@@ -0,0 +1,36 @@
+spec:
+  allowPrivilegeEscalation: true
+  allowedCapabilities:
+  - 'NET_ADMIN'
```
We only need this for cases where we install istio with sidecar injection enabled, right?

yu2003w replied:
Yes, serving activator and autoscaler need "NET_ADMIN" to startup.

mattmoor replied:
I think I'd rather have this in `test/config` until we figure out if/what we want to ship.

markusthoemmes replied:
@yu2003w I don't think it's the actual activator/autoscaler but rather the istio sidecars (or init containers in this case) that need these permissions.

yu2003w replied:
Sorry, I misunderstood that. It's Istio sidecar needs the permission.

yu2003w replied:
@mattmoor Moved psp.yaml to test/config.
When both psp and Istio sidecar automatically injection are enabled, deployments in namespace serving-tests also need 'NET_ADMIN'. So I also defined role and rolebinding within namespace serving-tests.

mattmoor commented on config/204-psp.yaml:
```diff
@@ -0,0 +1,36 @@
+  seLinux:
+    rule: RunAsAny
+  supplementalGroups:
+    rule: RunAsAny
```
Do these things need to be as permissive as they are given that we've stopped running as root?

yu2003w replied:
Because these fields are mandatory, without them, PSP couldn't created.
Without them following errors will be reported
```
error: error validating "psp.yaml": error validating data: [ValidationError(PodSecurityPolicy.spec): missing required field "seLinux" in io.k8s.api.extensions.v1beta1.PodSecurityPolicySpec, ValidationError(PodSecurityPolicy.spec): missing required field "runAsUser" in io.k8s.api.extensions.v1beta1.PodSecurityPolicySpec, ValidationError(PodSecurityPolicy.spec): missing required field "supplementalGroups" in io.k8s.api.extensions.v1beta1.PodSecurityPolicySpec, ValidationError(PodSecurityPolicy.spec): missing required field "fsGroup" in io.k8s.api.extensions.v1beta1.PodSecurityPolicySpec]; if you choose to ignore these errors, turn validation off with --validate=false
```

mattmoor replied:
I'm pretty new to PSP, but this looks like it allows us to do anything, so I'm curious how this is an improvement over where we were prior to `:nonroot`?

yu2003w replied:
Actually, if PSP is used, :nonroot images is not necessary any more.
Another choice is hand over PSP related stuff to cluster administrator to handle.

mattmoor replied:
Now I am _more_ confused :)

If PSP validates the `USER` then shouldn't it block our controllers, which run as root (without `:nonroot`)?

Maybe the answer is that I need to read what this does...

yu2003w replied:
If PSP validates user and container needs to run as root, it will be blocked.
For example, I created psp for Knative serving as below,
```
root@filmy1:/mnt/data/gowork/src/github.com/knative/serving# knative get psp knative-serving-psp
NAME                  PRIV   CAPS        SELINUX    RUNASUSER          FSGROUP    SUPGROUP   READONLYROOTFS   VOLUMES
knative-serving-psp   true   NET_ADMIN   RunAsAny   MustRunAsNonRoot   RunAsAny   RunAsAny   false            *
```
Then I build knative serving with base images ubuntu which need to run as root.
Pod failed to start as below,
```
  Normal   Pulling    3m34s                 kubelet, 172.16.17.35  pulling image "icpdev.icp:8500/imgrepo/controller-43f0364ab2f6dab17267e80a1f6a4adc@sha256:4806fc2cc89330d8731385033012f4f04ebc43d5f4918af999b7528162e9dd2b"
  Normal   Pulled     3m30s                 kubelet, 172.16.17.35  Successfully pulled image "icpdev.icp:8500/imgrepo/controller-43f0364ab2f6dab17267e80a1f6a4adc@sha256:4806fc2cc89330d8731385033012f4f04ebc43d5f4918af999b7528162e9dd2b"
  Warning  Failed     76s (x12 over 3m30s)  kubelet, 172.16.17.35  Error: container has runAsNonRoot and image will run as root
  Normal   Pulled     76s (x11 over 3m29s)  kubelet, 172.16.17.35  Container image "icpdev.icp:8500/imgrepo/controller-43f0364ab2f6dab17267e80a1f6a4adc@sha256:4806fc2cc89330d8731385033012f4f04ebc43d5f4918af999b7528162e9dd2b" already present on machine
```

yu2003w replied:
If Istio auto injection is enabled, it needs capability of 'NET_ADMIN' and needs to be run as root. And RunAsUser in psp should be set to RunAsAny.

mattmoor replied:
Ok, so `tl;dr` this is the PSP one needs with Istio Mesh, which is scary.

What I'm wondering is what **knative/serving** needs when run _without_ sidecars, e.g. istio-lean.yaml?

My hope was for us to be less scary (and use PSP to verify that), but I see little value in testing it in scenarios where the mesh potentially masks all of our work.  If we end up needing two PSPs (mesh and no-mesh) that's fine, and we can tweak scripts to install the right ones.

thanks for your patience while I wrap my head around this.

yu2003w commented:
/retest

yu2003w commented:
Currently no api could be used for determine which admission controller are enabled.
When such API available, if Istio and PodSecurityPolicy are detected, PSP should be created.
https://github.com/kubernetes/kubernetes/issues/79735

markusthoemmes commented:
@yu2003w right, so should we hide this behind a label so users only install this if they really need it?

On another note: This is kinda only related to istio and the way its configured. It's not really about Knative itself. Should this move into third_party and be documented accordingly?

/assign @mattmoor

mattmoor commented:
I have mixed feelings about this.  This is it's effectively testing Istio, not Knative, as it is oriented into such a permissive mode almost entirely driven by the injected sidecar.  I'm skeptical that it provides any meaningful coverage of Knative itself.

I actually like @markusthoemmes `third_party` suggestion best.  Could we attach this permissive PSP to the mesh version of istio and a more restrictive PSP to the lean install (that's more representative of Knative)?

markusthoemmes commented:
Not sure why this was moved to test/config @yu2003w. I think this **can** be useful in mesh-cases and if the user wants to install istio. As such, I think it can live under third-party/istio*

yu2003w commented:
> Not sure why this was moved to test/config @yu2003w. I think this **can** be useful in mesh-cases and if the user wants to install istio. As such, I think it can live under third-party/istio*

Sorry, I just read the comments.
At first, when I initiated this PR, we don't have a clear idea about how to handler PSP. I just added the object needed when PodSecurityPolicy is enabled. So I moved that to test/config.
It's a good idea to move that to thirdparty. I will update this PR.

mattmoor-sockpuppet commented on third_party/istio-psp.yaml:
```diff
@@ -0,0 +1,50 @@
+  - apiGroups: ["extensions"]
+    resources: ["podsecuritypolicies"]
+    verbs: ["use"]
+    resourceNames: ["knative-serving-psp"]
```
<!--whitespace-bot-->
```suggestion
    resourceNames: ["knative-serving-psp"]

```

yu2003w commented:
@markusthoemmes I move the PSP related stuff to thirdparty/istio-psp.yaml.
If PodSecurityPolicy is enabled, user could apply Istio-psp.yaml when install knative serving.

mattmoor commented:
Moving to `test/config` was so this got deployed when testing, with the latest round of changes we won't test this at all.  Honestly, I'm most interested in the "lean" PSP, which should be a truer reflection of the Knative PSP implications.

Can we add:
1. a "lean" PSP that accurately reflects the required PSP without a mesh
2. add logic to install the appropriate PSP alongside istio

yu2003w commented:
@mattmoor  I added back psp related files to test/config/psp. To run tests with PodSecurityPolicy enabled, user could run 'ko apply -f test/config -f test/config/psp' to generate necessary stuff.

For 'lean' psp, because 'NET_ADMIN' is required by Istio sidecar, RunAsUser should not be 'MustRunAsNonRoot'. 
After testing, I found that only volume types of 'configMap', 'emptyDir' and 'secret' are needed for istio sidecar. So istio-psp.yaml is updated.

Currently, maybe it's impossible to add logic to install appropriate psp along side with istio because it's difficult to determine which admission plugin are enabled when k8s api-server is started. [kubernetes#79735](https://github.com/kubernetes/kubernetes/issues/79735)

It's possibly psp couldn't installed alongside with istio until [kubernetes#79735](https://github.com/kubernetes/kubernetes/issues/79735) is resolved.

yu2003w commented:
/test pull-knative-serving-integration-tests

yu2003w commented:
/retest

mattmoor commented:
> Currently, maybe it's impossible to add logic to install appropriate psp along side with istio because it's difficult to determine which admission plugin are enabled when k8s api-server is started. kubernetes#79735

I understand that in general, however, we know which Istio each of our test setups install.  My hope was that we could configure the appropriate PSP with each of them vs. having a "lowest common demoninator" PSP that leaves a hole so big (due to the sidecar requirements) that you could drive a truck through it :(

Like I said, the most interesting PSP to me isn't `LCD(serving, istio w/ sidecar)` it is really something that paints an accurate picture of the *serving* requirements.  I think this is best captured by `LCD(serving, istio w/o sidecar)`.

Perhaps we could try to schedule a call to discuss this further because going back and forth here seems to be going in circles, and I'd like to make the best use of everyone's time.  Want to follow up in #networking or a group chat?

markusthoemmes commented:
Any updates? Been hanging for a good 3 months now with no activity.

markusthoemmes commented:
It hung another month. Closing for now, feel free to reopen if still valid.

/close
